### PR TITLE
Restrict email by client

### PR DESCRIPTION
**Pull request by JReissmueller**
The same client can now use the same email for multiple services.
Services with the same email now show/manage only their own licenses.
Abstracted code.
#15

### CODE REVIEW

File: cwatch.php [hunk 1]
```diff
@@ -189,116 +189,67 @@ public function addService(
...
-                    }
-
-                    // Determine what term to add these licenses for
-                    $license_term = 'MONTH_1';
-                    if (($pricing = $this->Services->getPackagePricing($vars['pricing_id']))) {
-                        $license_term = strtoupper($pricing->period) . '_' . $pricing->term;
-                    }
+        // Get a list of which email accounts are associated with this client
+        $module = $this->getModule();
+        $account_emails_meta = $this->ModuleClientMeta->get(
```
Comment (tysonphillips) [inline cwatch.php]: I imagine you might as well not make the call to ModuleClientMeta::get if you don't have a client ID.

File: cwatch.php [hunk 2]
```diff
@@ -321,147 +272,151 @@ public function addService(
...
-     * @param stdClass $service A stdClass object representing the current service
      * @param array $vars An array of user supplied info to satisfy the request
-     * @return array A numerically indexed array of meta fields to be stored for this service containing:
-     *  - key The key for this meta field
-     *  - value The value for this key
-     *  - encrypted Whether or not this field should be encrypted (default 0, not encrypted)
+     * @return array A list of licenses keys added
      */
-    private function updateService($service, array $vars = [])
+    public function pushUserAndLicenses($vars, array $service_licenses = null)
```
Comment (tysonphillips) [inline cwatch.php]: Does this method need to be public?

File: cwatch.php [hunk 3]
```diff
@@ -321,147 +272,151 @@ public function addService(
...
 
-                        $license = $license_response->response();
-                        $license_keys[] = $license->distributionResult[0]->licenseKeys[0];
+            $license_keys = [];
+            // Add licenses to the customer account according to the config options provided
+            foreach ($license_types as $license_type => $quantity) {
+                for ($i = 0; $i < $quantity; $i++) {
+                    $license_response = $api->addLicense(
+                        $license_type,
+                        $license_type == 'BASIC_DETECTION' ? 'UNLIMITED' : $license_term,
```
Comment (tysonphillips) [inline cwatch.php]: The terms like "BASIC_DETECTION" and "UNLIMITED" seem like they could be class constants, no?

File: cwatch.php [hunk 4]
```diff
@@ -321,147 +272,151 @@ public function addService(
...
+                    if ($license_response->status() != 200) {
+                        $this->Input->setErrors(['api' => ['internal' => $license_response->errors()]]);
+
+                        break 2;
                     }
+
+                    // Keep track of which licenses were created by this edit so they can be reverted or added to
+                    // the list of keys for this service
+                    $license = $license_response->response();
+                    $license_keys[] = $license->distributionResult[0]->licenseKeys[0];
```
Comment (tysonphillips) [inline cwatch.php:416]: Are you sure the license will always have a 0th index distributionResult and 0th index licenseKeys? I wouldn't want any "reference to non-object" or "array key does not exist" errors to be possible

File: cwatch.php [hunk 5]
```diff
@@ -189,116 +189,67 @@ public function addService(
...
+        if (isset($vars['use_module']) && $vars['use_module'] == 'true') {
+            // Add user and licenses
+            $license_keys = $this->pushUserAndLicenses($vars);
+
+            if ($this->Input->errors()) {
+                // Error
+                // Delete the user if this is the only service using this email
+                if (isset($vars['cwatch_email'])
+                    && !$account_emails_meta
+                    || (array_key_exists($vars['cwatch_email'], $account_emails_meta->value)
```
Comment (tysonphillips) [inline cwatch.php:224]: You may want to _strtolower_ the email address so you only store and compare email addresses that are lower case. For instance, we wouldn't want to add the same email address that only differs from a capital letter.

File: cwatch.php (hunk 4 above)
Comment (JReissmueller) [reply cwatch.php:416]: Assuming a successful response, which is checked for earlier, then yes this is the expected response format.

**Review by tysonphillips — APPROVED** [review]

**Pull request merged**